app/src/app/components/contents/runner/run/PacketRunBranchField.tsx

```diff
@@ -0,0 +1,91 @@
+import { Github, RefreshCw } from "lucide-react";
+import { Button } from "../../../Base/Button";
+import { FormControl, FormField, FormItem, FormLabel, FormMessage } from "../../../Base/Form";
+import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "../../../Base/Select";
+import { Separator } from "../../../Base/Separator";
+import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from "../../../Base/Tooltip";
+import { GitBranches, GitBranchInfo } from "../types/GitBranches";
+import { getTimeDifferenceToDisplay } from "../../explorer/utils/getTimeDifferenceToDisplay";
+import { KeyedMutator } from "swr";
+import { fetcher } from "../../../../../lib/fetch";
+import appConfig from "../../../../../config/appConfig";
+import { UseFormReturn } from "react-hook-form";
+import { z } from "zod";
+import { packetRunFormSchema } from "./PacketRunForm";
+
+interface PacketRunBranchFieldProps {
+  branches: GitBranchInfo[];
+  form: UseFormReturn<z.infer<typeof packetRunFormSchema>>;
+  mutate: KeyedMutator<GitBranches>;
+}
+export const PacketRunBranchField = ({ branches, form, mutate }: PacketRunBranchFieldProps) => {
+  const selectedBranch = branches.filter((branch) => branch.name === form.getValues("branch"))[0];
+  const lastCommitTime = getTimeDifferenceToDisplay(selectedBranch.time);
+
+  const gitFetch = async () => {
+    try {
+      await fetcher({
+        url: `${appConfig.apiUrl()}/runner/git/fetch`,
+        method: "POST"
+      });
+    } catch (error) {
+      console.error(error);
+    }
+    mutate();
+  };
+
+  return (
+    <>
+      <FormLabel className="font-semibold text-lg">Branch</FormLabel>
+      <div className="flex space-x-3">
+        <div className="flex-grow">
+          <FormField
+            control={form.control}
+            name="branch"
+            render={({ field }) => (
+              <FormItem>
+                <Select defaultValue={field.value} onValueChange={field.onChange}>
```

Comment:
Sorry, just to clarify as I'm still not super-familiar with this style of React forms - if I understand it, the field comes out of the  form which is passed as a prop into the component from the parent form, and it's selecting the field it wants to use from the "name" prop in `FormField` - and we're setting an `onChange` event handler here, which will trigger `onChange` on that field in the form, if it has one defined (which it doesn't currently - or is there a default handler implied by `useForm` which just updates the value?).

So it's a way to let components like this handle rendering of fields, but not needing to know anything about how to handle changes to values, as those event handlers are defined in whatever creates the 'form' object (the parent form component in this case)..?

Reply:
yip pretty much :smile